### ISSUE TITLE
Is possible to run script in RELEASE mode?

### ISSUE THREAD

**Issue by wk-j** [filipw/dotnet-script#221]
I'm trying to execute following code with dotnet script but it seem like BenchmarkDotNet not allow running benchmarks in DEBUG mode.

```csharp
#! "netcoreapp2.0"
#r "nuget:NetStandard.Library,2.0"
#r "nuget:BenchmarkDotNet,0.10.12"
#r "nuget:System.Memory,4.4.0-preview1-25305-02"

using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[MemoryDiagnoser]
public class Test {

    int[] values = new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 };

    [Benchmark]
    public void Take() {
        var a = values.Skip(5).Count();
    }

    [Benchmark]
    public void Slice() {
        var span = new Span<int>(values, 0);
        var a = span.Slice(5, span.Length - 5).Length;
    }
}

BenchmarkRunner.Run<Test>();
```

```
❯ dotnet script "./scripts/Slice.csx"
// ***** BenchmarkRunner: Start   *****
// Found benchmarks:
//   Test.Take: DefaultJob
//   Test.Slice: DefaultJob

// Validating benchmarks:
Assembly ℛ*beb7aab2-3d06-4c0c-83f8-7cd9cf431b19#2-0 which defines benchmarks is non-optimized
Benchmark was built without optimization enabled (most probably a DEBUG configuration). Please, build it in RELEASE.
// * Artifacts cleanup *
```

**Comment by seesharper** [comment]
H @wk-j 
There is currently no way to execute scripts in release mode. 
@filipw We should maybe add an option for this. It should be interesting to see if that affects the startup time. 

Something like 
```
-c, --configuration <CONFIGURATION>   Configuration to use for running the script [Release/Debug] Default is "Debug".
```

This way we can also add this to the `launch.json` file so that <kbd>CTRL</kbd>+<kbd>F5</kbd> (Start without debugging) adds the `-c release` option. That way we can execute in release mode from within VS Code

**Comment by filipw** [comment]
It was there, but we removed it. However, it didn't behave as expected anyway, so we might as well say it has never been there 😀

The reason for not having it, was really to align with CSI, which also doesn't have it, and emit symbols all the time. We can add it as an opt-in though.

**Comment by filipw** [comment]
Actually I just remembered that this is blocked by Roslyn (https://github.com/dotnet/roslyn/blob/master/src/Scripting/CSharp/CSharpScriptCompiler.cs#L57). I opened an issue 3 years ago 😂 https://github.com/dotnet/roslyn/issues/1262

**Comment by seesharper** [comment]
@filipw  Can we hack this with reflection?

**Comment by seesharper** [comment]
I got it working with release mode doing some reflection 

```
var compilation = script.GetCompilation();
var compilationOptionsField = typeof(CSharpCompilation).GetTypeInfo().GetDeclaredField("_options");
var compilationOptions = (CSharpCompilationOptions)compilationOptionsField.GetValue(compilation);
compilationOptions = compilationOptions.WithOptimizationLevel(OptimizationLevel.Release);
compilationOptionsField.SetValue(compilation, compilationOptions);
```

Problem is that BenchmarkDotNet chokes when it sees an in-memory assembly.

https://github.com/dotnet/BenchmarkDotNet/issues/377#issuecomment-362589626

**Comment by seesharper** [comment]
@filipw I might have a super simple approach to this.

```
if (Debugger.IsAttached)
{
    // Run in debug mode
}
else
{
  // Run in release mode.
}
```

This way we introduce no breaking changes with regards to existing `launch.json` files. and  will be able to execute the script within VS Code in release mode <kbd>CTRL</kbd>+<kbd>F5</kbd> (Start without debugging)

**Comment by filipw** [comment]
The issue would still be that then the debug info is missing, and because of that, the line numbers are not available in the stack traces, which is one of the most important things if I think about script authoring. 

So I really thing release mode should be opt-in.

**Comment by seesharper** [comment]
Good point. 👍 

What if we do this?
```
-c, --configuration <CONFIGURATION>   Configuration to use for running the script [Release/Debug] Default is "Debug".
```
if someone needs to run in release mode from VS Code they can always create a launch config for that purpose that adds the switch

**Comment by filipw** [comment]
yes this is exactly what I was thinking

**Comment by seesharper** [comment]
Sounds good, I'm on it 👍

**Comment by seesharper** [comment]
@filipw I've done some work on this and as it turns out the stack trace is not affected by the optimization level at all since we do emit debug symbols regardless of the optimization level (debug/release)

Test script:
```
#! "netcoreapp2.0"

Console.WriteLine("Hello world!");
throw new Exception();
```

Debug config:

```
$ dotnet dotnet-script.dll -c debug c:/scriptdemo/main.csx
Hello world!
System.Exception: Exception of type 'System.Exception' was thrown.
   at Submission#0.<<Initialize>>d__0.MoveNext() in c:\scriptdemo\main.csx:line 4
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.<RunSubmissionsAsync>d__9`1.MoveNext()
```

Release config 
```
$ dotnet dotnet-script.dll -c release c:/scriptdemo/main.csx
Hello world!
System.Exception: Exception of type 'System.Exception' was thrown.
   at Submission#0.<<Initialize>>d__0.MoveNext() in c:\scriptdemo\main.csx:line 4
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.<RunSubmissionsAsync>d__9`1.MoveNext()
```

Do we still want the switch or should we do the `Debugger.IsAttached` approach?

**Comment by filipw** [comment]
ahhh very nice. I need to think if there is anything else that would be affected but I don't think so. 
What if I had a long running script and wanted to attach the debugger to the process rather than start with debugger?

**Comment by seesharper** [comment]
Which is exactly what is happening here.
I'll forgot I write that 😄 
https://github.com/filipw/dotnet-script#debugging-1

Besides changing the default behaviour based on an enge case is never a good idea :)

I'll finish up with the switch leaving the default as it is 👍

**Comment by seesharper** [comment]
@wk-j Once we merge and release #229 you can execute the following script to benchmark

```
#! "netcoreapp2.0"
#r "nuget:System.Memory, 4.5.0-preview1-26216-02"
#r "nuget:BenchmarkDotNet,0.10.12"

using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Attributes.Jobs;

BenchmarkRunner.Run<Test>();

[InProcess]
public class Test
{       
    int[] values = new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 };

    [Benchmark]
    public void Take()
    {
        var a = values.Skip(5).Count();
    }

    [Benchmark]
    public void Slice()
    {
        var span = new Span<int>(values, 0, values.Length);
        var a = span.Slice(5, span.Length - 5).Length;
    }
}
```

specify release configuration like this 

```
dotnet script -c release benchmark.csx
```

**Comment by seesharper** [comment]
Fixed by #229
